Build details page: normalize/trim command paths (second attempt)

humitos opened the pull request:
Continuation of the work done in
#9815 that didn't work as we
expected.

We weren't saving the commands properly when hitting the API since
`serializers.SerializerMethodField` is a read-only field. This commit fixes this
problem by using a different Serializer depending if it's a GET request or not,
and if it's made by an Admin or not.

It overrides the method `BuildViewSet.get_serializer_class` to manage these
different cases and return the proper serializer.

Done on top of #9827 to be sure the
tests for saving commands are passing.

Closes #9833

stsewd commented on readthedocs/api/v2/serializers.py:
```diff
@@ -133,11 +136,49 @@ class Meta:
         exclude = []
 
 
+class BuildCommandUISerializer(BuildCommandSerializer):
```
I would name these `Read`/`ReadOnly`, since it isn't necessarily just the UI that use these, but we use them when listing/retrieving a record

```suggestion
class BuildCommandReadSerializer(BuildCommandSerializer):
```

stsewd reviewed (commented):
Should this also be implemented for the command API? https://readthedocs.org/api/v2/command/174704149/

Don't think we use that from our dashboard, but it will make the API consistent.

ericholscher commented on readthedocs/api/v2/serializers.py:
```diff
@@ -162,11 +203,17 @@ class BuildAdminSerializer(BuildSerializer):
 
 
+class BuildAdminUISerializer(BuildAdminSerializer):
+    commands = BuildCommandUISerializer(many=True, read_only=True)
```
This needs a docstring, I'm pretty confused about all the different serializers.

ericholscher commented on readthedocs/api/v2/serializers.py:
```diff
@@ -133,11 +136,49 @@ class Meta:
     """Build serializer for user display, doesn't display internal fields."""
 
-    commands = BuildCommandSerializer(many=True, read_only=True)
+    commands = BuildCommandUISerializer(many=True, read_only=True)
```
We definitely need a comment or something explaining why this defaults to the UI serializer, instead of the regular one? It seems confusing to me we have:

BuildSerializer -> BuildAdminSerializer -> BuildAdminUISerializer

Why do we have 3 levels of nesting here, and what are each used for? It's not clear from the docstrings.

ericholscher commented on readthedocs/api/v2/views/model_views.py:
```diff
@@ -224,11 +225,25 @@ class VersionViewSet(DisableListEndpoint, UserSelectViewSet):
+            # Logic copied from `UserSelectViewSet.get_serializer_class`
+            # and extended to check for GET method
+            if self.request.method == "GET":
+                return BuildAdminUISerializer
+            return BuildAdminSerializer
+        return BuildSerializer
```
```suggestion
                return BuildAdminUISerializer #Staff read-only
            return BuildAdminSerializer # Staff write-only
        return BuildSerializer # Non-staff
```

There's probably a way to make this logic more clear, but the way it is now could definitely use some clarification.

ericholscher reviewed (commented):
This seems like it will fix the issue, along with @stsewd's feedback 👍

humitos commented:
@stsewd 

> Should this also be implemented for the command API?
> Don't think we use that from our dashboard, but it will make the API consistent.

I wasn't sure about this since we are not using it anywhere AFAIK, APIv2 is already deprecated and users shouldn't be using it. The main idea was to make the dashboard UI better for now without going too deep. If we want to follow that road, we should do it for APIv3 as well (but it will change the behavior, so 🤷🏼 ) and review all the places where they are used.

I think it may be fine to not implement it on the command API for now; but I'm open to other arguments.

humitos commented:
Thanks for the feedback! I think I've addressed all of it. Let me know.

@stsewd I renamed the serializer class to mention "ReadOnly" on its name.

@ericholscher I added docstring to the classes --hopefully, it's a little more clear now.

benjaoming commented on readthedocs/api/v2/serializers.py:
```diff
@@ -133,11 +136,63 @@ class Meta:
+class BuildCommandReadOnlySerializer(BuildCommandSerializer):
+
+    """
+    Serializer used on GETs to trimm the commands' path.
```
```suggestion
    Serializer used on GETs to trim the commands' path.
```

I trimmed trimm to trim :)

benjaoming commented on readthedocs/api/v2/views/model_views.py:
```diff
@@ -224,11 +225,25 @@ class VersionViewSet(DisableListEndpoint, UserSelectViewSet):
+        also want to have a specific serializer for admins.
+        """
+        if self.request.user.is_staff:
+            # Logic copied from `UserSelectViewSet.get_serializer_class`
+            # and extended to check for GET method
+            if self.action in ["list", "retrieve"]:
+                return BuildAdminReadOnlySerializer  # Staff read-onlyl
+            return BuildAdminSerializer  # Staff write-only
```
```suggestion
            # Logic copied from `UserSelectViewSet.get_serializer_class`
            # and extended to choose serializer from self.action
            if self.action not in ["list", "retrieve"]:
                return BuildAdminSerializer  # Staff write-only
            return BuildAdminReadOnlySerializer  # Staff read-only
```

This is a really small thing, but since the `BuildAdminSerializer` is the one with the elevated permissions, the conditions are better to set explicitly.

(but `self.request.user.is_staff` does the job very well otherwise)

benjaoming reviewed (approved):
Would be great to have this merged, and all the other feedback is resolved, so :+1: great work on adding comments!

ericholscher commented:
> @ericholscher I added docstring to the classes --hopefully, it's a little more clear now.

Much clearer what is happening!  💯
